File: src/vct/col/ast/ConstantExpression.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
package vct.col.ast;

import hre.ast.Origin;

import java.util.*;

import vct.col.ast.PrimitiveType.Sort;

/**
 * AST node for wrapping constant values.
 * 
 * @author sccblom
 *
 */
public class ConstantExpression extends ASTNode {

  public final Value value;

  public ConstantExpression(int i,Origin origin){
    this.value=new IntegerValue(i);
    setType(new PrimitiveType(Sort.Integer));
    setOrigin(origin);
  }
  public ConstantExpression(int i){
    this.value=new IntegerValue(i);
    setType(new PrimitiveType(Sort.Integer));
  }
  
  public ConstantExpression(Value v,Type t,Origin origin){
    this.value=v;
    setType(t);
    setOrigin(origin);
  }

  public ConstantExpression(boolean b,Origin origin){
    this.value=new BooleanValue(b);
    setType(new PrimitiveType(Sort.Boolean));
    setOrigin(origin);
  }
  public ConstantExpression(boolean b){
    this.value=new BooleanValue(b);
    setType(new PrimitiveType(Sort.Boolean));
  }

  public ConstantExpression(String s, Origin origin) {
    this.value=new StringValue(s);
    setType(new PrimitiveType(Sort.String));
    setOrigin(origin);
  }
  public ConstantExpression(String s){
    this.value=new StringValue(s);
    setType(new PrimitiveType(Sort.String));
  }

  public ConstantExpression(long i, Origin origin) {
    this.value=new LongValue(i);
    setType(new PrimitiveType(Sort.Long));
    setOrigin(origin);
  }
  public ConstantExpression(long i) {
    this.value=new LongValue(i);
    setType(new PrimitiveType(Sort.Long));
  }

  public ConstantExpression(double d, Origin origin) {
    this.value=new DoubleValue(d);
    setType(new PrimitiveType(Sort.Double));
    setOrigin(origin);
  }
  public ConstantExpression(double d) {
    this.value=new DoubleValue(d);
    setType(new PrimitiveType(Sort.Double));
  }
  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    visitor.visit(this);
  }
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    return map.map(this);
  }

  public Value getValue(){
    return value;
  }
  public String toString(){
    return value.toString();
  }
  
  public boolean equals(Object o){
    return value.equals(o);
  }
}

File: vercors/src/main/java/vct/col/ast/expr/constant/ConstantExpression.scala
package vct.col.ast.expr.constant

import java.lang

import hre.ast.Origin
import vct.col.ast.`type`.{PrimitiveSort, PrimitiveType, Type}
import vct.col.ast.generic.ASTNode
import vct.col.ast.stmt.composite.Hole
import vct.col.ast.util.ASTVisitor
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}
import collection.JavaConverters._

import scala.collection.JavaConverters._

/**
 * AST node for wrapping constant values, e.g. integers, booleans, strings, doubles, and longs.
 *
 * @author sccblom, whmoortwijn
 * @param value The constant value that is wrapped by this node.
 */
case class ConstantExpression(val value:Value) extends ASTNode with VisitorHelper {
  def this(v:Value, t:Type) = { this(v); setType(t) }
  def this(v:Value, t:Type, origin:Origin) = { this(v, t); setOrigin(origin) }
  def this(i:Int) = this(new IntegerValue(i), new PrimitiveType(PrimitiveSort.Integer))
  def this(b:Boolean) = this(new BooleanValue(b), new PrimitiveType(PrimitiveSort.Boolean))
  def this(s:String) = this(new StringValue(s), new PrimitiveType(PrimitiveSort.String))
  def this(l:Long) = this(new LongValue(l), new PrimitiveType(PrimitiveSort.Long))
	def this(d:Double) = this(new DoubleValue(d), new PrimitiveType(PrimitiveSort.Double))

  def this(i:Int, origin:Origin) = { this(i); setOrigin(origin) }
  def this(b:Boolean, origin:Origin) = { this(b); setOrigin(origin) }
  def this(s:String, origin:Origin) = { this(s); setOrigin(origin) }
  def this(l:Long, origin:Origin) = { this(l); setOrigin(origin) }
  def this(d:Double, origin:Origin) = { this(d); setOrigin(origin) }

  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def equals(o:Any) : Boolean = value.equals(o)
  override def isConstant(o:Any) : Boolean = equals(o)
  override def toString() : String = value.toString()

  override def `match`(ast:ASTNode) = ast match {
    case h:Hole => h.`match`(this)
    case _ => ast.equals(value)
  }

  override def debugTreeChildrenFields(): Iterable[String] = Seq()
  override def debugTreePropertyFields(): Iterable[String] = Seq("value")
}
